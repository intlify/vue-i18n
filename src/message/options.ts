import { CompileError } from './errors'

export type CompileErrorHandler = (error: CompileError) => void
export type CompileCacheKeyHandler = (source: string) => string

export type TokenizeOptions = {
  // TODO: other options
  location?: boolean
  onError?: CompileErrorHandler
}

export type ParserOptions = {
  // TODO: other options
  // Generate source map?
  // - Default: false
  // sourceMap?: boolean
  // Filename for source map generation.
  // - Default: `message.intl`
  // filename?: string
  location?: boolean
  onError?: CompileErrorHandler
}

export type TransformOptions = {
  // TODO: other options
  // onError?: CompileErrorHandler
  onError?: CompileErrorHandler
}

export type CodeGenOptions = {
  mode?: 'normal' | 'arrow' // default normal
  onError?: CompileErrorHandler
}

export type CompileOptions = {
  warnHtmlMessage?: boolean
  onCacheKey?: CompileCacheKeyHandler
} & TransformOptions &
  CodeGenOptions &
  ParserOptions &
  TokenizeOptions
