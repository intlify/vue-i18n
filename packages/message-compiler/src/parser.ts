import { Position, createLocation, SourceLocation } from './location'
import { ParserOptions } from './options'
import { createCompileError, CompileErrorCodes } from './errors'
import { Tokenizer, createTokenizer, TokenTypes, Token } from './tokenizer'
import { assign } from '@intlify/shared'

export const enum NodeTypes {
  Resource, // 0
  Plural,
  Message,
  Text,
  Named,
  List, // 5
  Linked,
  LinkedKey,
  LinkedModifier,
  Literal
}

// not containing whitespace or control characters
export type Identifier = string

export interface Node {
  type: NodeTypes
  start: number
  end: number
  loc?: SourceLocation
}

export interface ResourceNode extends Node {
  type: NodeTypes.Resource
  body: MessageNode | PluralNode
  helpers?: string[]
}

export interface PluralNode extends Node {
  type: NodeTypes.Plural
  cases: MessageNode[]
}

export interface MessageNode extends Node {
  type: NodeTypes.Message
  items: MessageElementNode[]
}

type MessageElementNode =
  | TextNode
  | NamedNode
  | ListNode
  | LiteralNode
  | LinkedNode

export interface TextNode extends Node {
  type: NodeTypes.Text
  value: string
}

export interface NamedNode extends Node {
  type: NodeTypes.Named
  key: Identifier
}

export interface ListNode extends Node {
  type: NodeTypes.List
  index: number
}

export interface LiteralNode extends Node {
  type: NodeTypes.Literal
  value: string
}

export interface LinkedNode extends Node {
  type: NodeTypes.Linked
  modifier?: LinkedModifierNode
  key: LinkedKeyNode | NamedNode | ListNode | LiteralNode
}

export interface LinkedKeyNode extends Node {
  type: NodeTypes.LinkedKey
  value: string
}

export interface LinkedModifierNode extends Node {
  type: NodeTypes.LinkedModifier
  value: Identifier
}

export interface Parser {
  parse(source: string): ResourceNode
}

export const ERROR_DOMAIN = 'parser'

// Backslash backslash, backslash quote, uHHHH, UHHHHHH.
const KNOWN_ESCAPES = /(?:\\\\|\\'|\\u([0-9a-fA-F]{4})|\\U([0-9a-fA-F]{6}))/g

function fromEscapeSequence(
  match: string,
  codePoint4: string,
  codePoint6: string
): string {
  switch (match) {
    case `\\\\`:
      return `\\`
    case `\\\'`:
      return `\'`
    default: {
      const codePoint = parseInt(codePoint4 || codePoint6, 16)
      if (codePoint <= 0xd7ff || codePoint >= 0xe000) {
        return String.fromCodePoint(codePoint)
      }
      // invalid ...
      // Replace them with U+FFFD REPLACEMENT CHARACTER.
      return 'ï¿½'
    }
  }
}

export function createParser(options: ParserOptions = {}): Parser {
  const location = options.location !== false

  const { onError } = options
  function emitError(
    tokenzer: Tokenizer,
    code: CompileErrorCodes,
    start: Position,
    offset: number,
    ...args: unknown[]
  ): void {
    const end = tokenzer.currentPosition()
    end.offset += offset
    end.column += offset
    if (onError) {
      const loc = createLocation(start, end)
      const err = createCompileError(code, loc, {
        domain: ERROR_DOMAIN,
        args
      })
      onError(err)
    }
  }

  function startNode(type: NodeTypes, offset: number, loc: Position): Node {
    const node = {
      type,
      start: offset,
      end: offset
    } as Node

    if (location) {
      node.loc = { start: loc, end: loc }
    }

    return node
  }

  function endNode(
    node: Node,
    offset: number,
    pos: Position,
    type?: NodeTypes
  ): void {
    node.end = offset

    if (type) {
      node.type = type
    }

    if (location && node.loc) {
      node.loc.end = pos
    }
  }

  function parseText(tokenizer: Tokenizer, value: string): TextNode {
    const context = tokenizer.context()
    const node = startNode(
      NodeTypes.Text,
      context.offset,
      context.startLoc
    ) as TextNode
    node.value = value
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseList(tokenizer: Tokenizer, index: string): ListNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.List, offset, loc) as ListNode
    node.index = parseInt(index, 10)
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseNamed(tokenizer: Tokenizer, key: string): NamedNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.Named, offset, loc) as NamedNode
    node.key = key
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLiteral(tokenizer: Tokenizer, value: string): LiteralNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.Literal, offset, loc) as LiteralNode
    node.value = value.replace(KNOWN_ESCAPES, fromEscapeSequence)
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLinkedModifier(
    tokenizer: Tokenizer
  ): {
    nextConsumeToken?: Token
    node: LinkedModifierNode
  } {
    const token = tokenizer.nextToken()
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get linked dot loc
    const node = startNode(
      NodeTypes.LinkedModifier,
      offset,
      loc
    ) as LinkedModifierNode
    if (token.type !== TokenTypes.LinkedModifier) {
      // empty modifier
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_MODIFIER,
        context.lastStartLoc,
        0
      )
      node.value = ''
      endNode(node, offset, loc)
      return {
        nextConsumeToken: token,
        node
      }
    }
    // check token
    if (token.value == null) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        getTokenCaption(token)
      )
    }
    node.value = token.value || ''
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return {
      node
    }
  }

  function parseLinkedKey(tokenizer: Tokenizer, value: string): LinkedKeyNode {
    const context = tokenizer.context()
    const node = startNode(
      NodeTypes.LinkedKey,
      context.offset,
      context.startLoc
    ) as LinkedKeyNode
    node.value = value
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLinked(
    tokenizer: Tokenizer
  ): {
    nextConsumeToken?: Token
    node: LinkedNode
  } {
    const context = tokenizer.context()
    const linkedNode = startNode(
      NodeTypes.Linked,
      context.offset,
      context.startLoc
    ) as LinkedNode

    let token = tokenizer.nextToken()
    if (token.type === TokenTypes.LinkedDot) {
      const parsed = parseLinkedModifier(tokenizer)
      linkedNode.modifier = parsed.node
      token = parsed.nextConsumeToken || tokenizer.nextToken()
    }

    // asset check token
    if (token.type !== TokenTypes.LinkedDelimiter) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        getTokenCaption(token)
      )
    }
    token = tokenizer.nextToken()

    // skip brace left
    if (token.type === TokenTypes.BraceLeft) {
      token = tokenizer.nextToken()
    }

    switch (token.type) {
      case TokenTypes.LinkedKey:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseLinkedKey(tokenizer, token.value || '')
        break
      case TokenTypes.Named:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseNamed(tokenizer, token.value || '')
        break
      case TokenTypes.List:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseList(tokenizer, token.value || '')
        break
      case TokenTypes.Literal:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseLiteral(tokenizer, token.value || '')
        break
      default:
        // empty key
        emitError(
          tokenizer,
          CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_KEY,
          context.lastStartLoc,
          0
        )
        const nextContext = tokenizer.context()
        const emptyLinkedKeyNode = startNode(
          NodeTypes.LinkedKey,
          nextContext.offset,
          nextContext.startLoc
        ) as LinkedKeyNode
        emptyLinkedKeyNode.value = ''
        endNode(emptyLinkedKeyNode, nextContext.offset, nextContext.startLoc)
        linkedNode.key = emptyLinkedKeyNode
        endNode(linkedNode, nextContext.offset, nextContext.startLoc)
        return {
          nextConsumeToken: token,
          node: linkedNode
        }
        break
    }

    endNode(linkedNode, tokenizer.currentOffset(), tokenizer.currentPosition())
    return {
      node: linkedNode
    }
  }

  function parseMessage(tokenizer: Tokenizer): MessageNode {
    const context = tokenizer.context()
    const startOffset =
      context.currentType === TokenTypes.Pipe
        ? tokenizer.currentOffset()
        : context.offset
    const startLoc =
      context.currentType === TokenTypes.Pipe
        ? context.endLoc
        : context.startLoc
    const node = startNode(
      NodeTypes.Message,
      startOffset,
      startLoc
    ) as MessageNode
    node.items = []

    let nextToken: Token | null = null
    do {
      const token = nextToken || tokenizer.nextToken()
      nextToken = null
      switch (token.type) {
        case TokenTypes.Text:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseText(tokenizer, token.value || ''))
          break
        case TokenTypes.List:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseList(tokenizer, token.value || ''))
          break
        case TokenTypes.Named:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseNamed(tokenizer, token.value || ''))
          break
        case TokenTypes.Literal:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseLiteral(tokenizer, token.value || ''))
          break
        case TokenTypes.LinkedAlias:
          const parsed = parseLinked(tokenizer)
          node.items.push(parsed.node)
          nextToken = parsed.nextConsumeToken || null
          break
      }
    } while (
      context.currentType !== TokenTypes.EOF &&
      context.currentType !== TokenTypes.Pipe
    )

    // adjust message node loc
    const endOffset =
      context.currentType === TokenTypes.Pipe
        ? context.lastOffset
        : tokenizer.currentOffset()
    const endLoc =
      context.currentType === TokenTypes.Pipe
        ? context.lastEndLoc
        : tokenizer.currentPosition()

    endNode(node, endOffset, endLoc)
    return node
  }

  function parsePlural(
    tokenizer: Tokenizer,
    offset: number,
    loc: Position,
    msgNode: MessageNode
  ): PluralNode {
    const context = tokenizer.context()
    let hasEmptyMessage = msgNode.items.length === 0

    const node = startNode(NodeTypes.Plural, offset, loc) as PluralNode
    node.cases = []
    node.cases.push(msgNode)

    do {
      const msg = parseMessage(tokenizer)
      if (!hasEmptyMessage) {
        hasEmptyMessage = msg.items.length === 0
      }
      node.cases.push(msg)
    } while (context.currentType !== TokenTypes.EOF)

    if (hasEmptyMessage) {
      emitError(
        tokenizer,
        CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL,
        loc,
        0
      )
    }

    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseResource(tokenizer: Tokenizer): MessageNode | PluralNode {
    const context = tokenizer.context()
    const { offset, startLoc } = context

    const msgNode = parseMessage(tokenizer)
    if (context.currentType === TokenTypes.EOF) {
      return msgNode
    } else {
      return parsePlural(tokenizer, offset, startLoc, msgNode)
    }
  }

  function parse(source: string): ResourceNode {
    const tokenizer = createTokenizer(source, assign({}, options))
    const context = tokenizer.context()

    const node = startNode(
      NodeTypes.Resource,
      context.offset,
      context.startLoc
    ) as ResourceNode
    if (location && node.loc) {
      node.loc.source = source
    }
    node.body = parseResource(tokenizer)

    // assert whether achieved to EOF
    if (context.currentType !== TokenTypes.EOF) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        source[context.offset] || ''
      )
    }

    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  return { parse }
}

function getTokenCaption(token: Token) {
  if (token.type === TokenTypes.EOF) {
    return 'EOF'
  }
  const name = (token.value || '').replace(/\r?\n/gu, '\\n')
  return name.length > 10 ? name.slice(0, 9) + 'â¦' : name
}
