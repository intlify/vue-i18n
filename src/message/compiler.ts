import { CompileOptions } from './options'
import { ResourceNode, createParser } from './parser'
import { transform } from './transformer'
import { generate } from './generator'
import { CompileError, defaultOnError } from './errors'
import { MessageFunction, MessageFunctions } from './runtime'
import { warn, format, isBoolean } from '../utils'

export interface CompileResult {
  code: string
  ast: ResourceNode
  // TODO: should be implemetend sourcemap
  // map?: RawSourceMap
}

const RE_HTML_TAG = /<\/?[\w\s="/.':;#-\/]+>/
const WARN_MESSAGE = `Detected HTML in '{source}' message. Recommend not using HTML messages to avoid XSS.`

function checkHtmlMessage(source: string, options: CompileOptions): void {
  const warnHtmlMessage = isBoolean(options.warnHtmlMessage)
    ? options.warnHtmlMessage
    : true
  if (warnHtmlMessage && RE_HTML_TAG.test(source)) {
    warn(format(WARN_MESSAGE, { source }))
  }
}

const defaultOnCacheKey = (source: string): string => source
let compileCache: unknown = Object.create(null)

export function clearCompileCache(): void {
  compileCache = Object.create(null)
}

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CompileResult {
  // parse source codes
  const parser = createParser({ ...options })
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, { ...options })

  // generate javascript codes
  const code = generate(ast, { ...options })

  return { ast, code }
}

export function compile<T = string>(
  source: string,
  options: CompileOptions = {}
): MessageFunction<T> {
  // check HTML message
  __DEV__ && checkHtmlMessage(source, options)

  // check caches
  const onCacheKey = options.onCacheKey || defaultOnCacheKey
  const key = onCacheKey(source)
  const cached = (compileCache as MessageFunctions<T>)[key]
  if (cached) {
    return cached
  }

  // compile error detecting
  let occured = false
  const onError = options.onError || defaultOnError
  options.onError = (err: CompileError): void => {
    occured = true
    onError(err)
  }

  // compile
  const { code } = baseCompile(source, options)

  // evaluate function
  const msg = new Function(`return ${code}`)() as MessageFunction<T>

  // if occured compile error, don't cache
  return !occured ? ((compileCache as MessageFunctions<T>)[key] = msg) : msg
}
