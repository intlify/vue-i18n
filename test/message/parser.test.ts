import { createParser } from '../../src/message/parser'

test('parse', () => {
  const parser = createParser()
  expect(parser.parse('hello world')).toMatchSnapshot()
  expect(parser.parse('hello {name} !')).toMatchSnapshot('named')
  expect(parser.parse('hello {0} !')).toMatchSnapshot('list')
  expect(parser.parse(`hello {'kazupon'} !`)).toMatchSnapshot('literal')
  expect(parser.parse('@:apples')).toMatchSnapshot('linked')
  expect(parser.parse('@.lower:{0}')).toMatchSnapshot('linked modifier list')
  expect(
    parser.parse('no apples | one apple  |  too much apples  ')
  ).toMatchSnapshot('plural')
  expect(
    parser.parse(`@.lower:{'no apples'} | {1} apple | {count}ã€€apples`) // eslint-disable-line no-irregular-whitespace
  ).toMatchSnapshot('plural complex')
})

/*
describe('Text', () => {
  test('simple', () => {
    const parser = createParser()
    expect(parser.parse('hello world')).toMatchSnapshot()
  })

  test('included end brace', () => {
    const spy = jest.fn()
    const parser = createParser({ onError: spy })
    expect(parser.parse('hello :-}')).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()
  })

  test('included end paren', () => {
    const spy = jest.fn()
    const parser = createParser({ onError: spy })
    expect(parser.parse('hello :-)')).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()
  })

  test('include at sign', () => {
    const spy = jest.fn()
    const parser = createParser({ onError: spy })
    expect(parser.parse('foo@bar.com')).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()
  })
})
*/
