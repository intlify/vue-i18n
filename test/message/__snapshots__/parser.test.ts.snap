// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 11,
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello world",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked 1`] = `
Object {
  "body": Object {
    "end": 8,
    "items": Array [
      Object {
        "end": 8,
        "key": Object {
          "end": 8,
          "loc": Object {
            "end": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "start": 2,
          "type": 7,
          "value": "apples",
        },
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 6,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 8,
  "loc": Object {
    "end": Object {
      "column": 9,
      "line": 1,
      "offset": 8,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked modifier list 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 11,
        "key": Object {
          "end": 11,
          "index": 0,
          "loc": Object {
            "end": Object {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
          },
          "start": 8,
          "type": 5,
        },
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "modifier": Object {
          "end": 7,
          "loc": Object {
            "end": Object {
              "column": 8,
              "line": 1,
              "offset": 7,
            },
            "start": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
          },
          "start": 1,
          "type": 8,
          "value": "lower",
        },
        "start": 0,
        "type": 6,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: list 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 6,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello ",
      },
      Object {
        "end": 9,
        "index": 0,
        "loc": Object {
          "end": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 5,
      },
      Object {
        "end": 11,
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
        },
        "start": 9,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: literal 1`] = `
Object {
  "body": Object {
    "end": 19,
    "items": Array [
      Object {
        "end": 6,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello ",
      },
      Object {
        "end": 17,
        "loc": Object {
          "end": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 9,
        "value": "kazupon",
      },
      Object {
        "end": 19,
        "loc": Object {
          "end": Object {
            "column": 20,
            "line": 1,
            "offset": 19,
          },
          "start": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
        },
        "start": 17,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 20,
        "line": 1,
        "offset": 19,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 19,
  "loc": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: named 1`] = `
Object {
  "body": Object {
    "end": 14,
    "items": Array [
      Object {
        "end": 6,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello ",
      },
      Object {
        "end": 12,
        "key": "name",
        "loc": Object {
          "end": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 4,
      },
      Object {
        "end": 14,
        "loc": Object {
          "end": Object {
            "column": 15,
            "line": 1,
            "offset": 14,
          },
          "start": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
        },
        "start": 12,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 14,
  "loc": Object {
    "end": Object {
      "column": 15,
      "line": 1,
      "offset": 14,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: plural 1`] = `
Object {
  "body": Object {
    "cases": Array [
      Object {
        "end": 0,
        "items": Array [
          Object {
            "end": 9,
            "loc": Object {
              "end": Object {
                "column": 10,
                "line": 1,
                "offset": 9,
              },
              "start": Object {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "start": 0,
            "type": 3,
            "value": "no apples",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      Object {
        "end": 12,
        "items": Array [
          Object {
            "end": 21,
            "loc": Object {
              "end": Object {
                "column": 22,
                "line": 1,
                "offset": 21,
              },
              "start": Object {
                "column": 13,
                "line": 1,
                "offset": 12,
              },
            },
            "start": 12,
            "type": 3,
            "value": "one apple",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
          "start": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
        },
        "start": 12,
        "type": 2,
      },
      Object {
        "end": 43,
        "items": Array [
          Object {
            "end": 43,
            "loc": Object {
              "end": Object {
                "column": 44,
                "line": 1,
                "offset": 43,
              },
              "start": Object {
                "column": 27,
                "line": 1,
                "offset": 26,
              },
            },
            "start": 26,
            "type": 3,
            "value": "too much apples  ",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 44,
            "line": 1,
            "offset": 43,
          },
          "start": Object {
            "column": 27,
            "line": 1,
            "offset": 26,
          },
        },
        "start": 26,
        "type": 2,
      },
    ],
    "end": 43,
    "loc": Object {
      "end": Object {
        "column": 44,
        "line": 1,
        "offset": 43,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 43,
  "loc": Object {
    "end": Object {
      "column": 44,
      "line": 1,
      "offset": 43,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: plural complex 1`] = `
Object {
  "body": Object {
    "cases": Array [
      Object {
        "end": 18,
        "items": Array [
          Object {
            "end": 19,
            "key": Object {
              "end": 18,
              "loc": Object {
                "end": Object {
                  "column": 19,
                  "line": 1,
                  "offset": 18,
                },
                "start": Object {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
              },
              "start": 9,
              "type": 7,
              "value": "no apples",
            },
            "loc": Object {
              "end": Object {
                "column": 20,
                "line": 1,
                "offset": 19,
              },
              "start": Object {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "modifier": Object {
              "end": 7,
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
                "start": Object {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "start": 1,
              "type": 8,
              "value": "lower",
            },
            "start": 0,
            "type": 6,
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 20,
            "line": 1,
            "offset": 19,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      Object {
        "end": 25,
        "items": Array [
          Object {
            "end": 25,
            "index": 1,
            "loc": Object {
              "end": Object {
                "column": 26,
                "line": 1,
                "offset": 25,
              },
              "start": Object {
                "column": 23,
                "line": 1,
                "offset": 22,
              },
            },
            "start": 22,
            "type": 5,
          },
          Object {
            "end": 31,
            "loc": Object {
              "end": Object {
                "column": 32,
                "line": 1,
                "offset": 31,
              },
              "start": Object {
                "column": 26,
                "line": 1,
                "offset": 25,
              },
            },
            "start": 25,
            "type": 3,
            "value": " apple",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 32,
            "line": 1,
            "offset": 31,
          },
          "start": Object {
            "column": 23,
            "line": 1,
            "offset": 22,
          },
        },
        "start": 22,
        "type": 2,
      },
      Object {
        "end": 48,
        "items": Array [
          Object {
            "end": 41,
            "key": "count",
            "loc": Object {
              "end": Object {
                "column": 42,
                "line": 1,
                "offset": 41,
              },
              "start": Object {
                "column": 35,
                "line": 1,
                "offset": 34,
              },
            },
            "start": 34,
            "type": 4,
          },
          Object {
            "end": 48,
            "loc": Object {
              "end": Object {
                "column": 49,
                "line": 1,
                "offset": 48,
              },
              "start": Object {
                "column": 42,
                "line": 1,
                "offset": 41,
              },
            },
            "start": 41,
            "type": 3,
            "value": "　apples",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 49,
            "line": 1,
            "offset": 48,
          },
          "start": Object {
            "column": 35,
            "line": 1,
            "offset": 34,
          },
        },
        "start": 34,
        "type": 2,
      },
    ],
    "end": 48,
    "loc": Object {
      "end": Object {
        "column": 49,
        "line": 1,
        "offset": 48,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 48,
  "loc": Object {
    "end": Object {
      "column": 49,
      "line": 1,
      "offset": 48,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;
